stl/inc/memory

```diff
@@ -4000,15 +4000,18 @@ protected:
     _Atomic_ptr_base(remove_extent_t<_Ty>* const _Px, _Ref_count_base* const _Ref) noexcept
         : _Ptr(_Px), _Repptr(_Ref) {}
 
-    void _Wait(remove_extent_t<_Ty>* _Old, memory_order) const noexcept {
+    void _Wait(remove_extent_t<_Ty>* _Old_ptr, _Ref_count_base* const _Old_rep, memory_order) const noexcept {
+        unsigned long _Remaining_timeout = 16; // milliseconds
+        const unsigned long _Max_timeout = 1048576; // milliseconds, ~17.5 minutes
         for (;;) {
             auto _Rep   = _Repptr._Lock_and_load();
-            bool _Equal = _Ptr.load(memory_order_relaxed) == _Old;
+            bool _Equal = _Ptr.load(memory_order_relaxed) == _Old_ptr && _Rep == _Old_rep;

```

Comment:
No change requested; we should do this in a followup if at all to avoid resetting testing, since it's not a correctness issue. Should this be reordered as:
```suggestion
            bool _Equal = _Rep == _Old_rep && _Ptr.load(memory_order_relaxed) == _Old_ptr;
```
so we short-circuit when the core-local `_Rep == _Old_rep` is `false` before potentially loading `_Ptr`'s cache line from some other core's data cache? (I suspect I understand cache coherence protocols just well enough to be dangerous.) Would any difference just be noise compared to the expense of `_Repptr._Lock_and_load()`?

Reply:
(Pulling in @AlexGuteniev to render an opinion and/or tell me how wrong I am 😄.)

Reply:
My vague understanding is that a relaxed load has no barriers, so it has no special costs. It would be fine to reorder, though, since the `.load` is at least a debug mode function call.

Reply:
Aren't _Repptr and _Ptr in the same cache line? Sure, whatever they point to can be anywhere, but the pointers themselves are (as far as I can see) right beside each other.

(STL's comment sounds accurate, though.)

Reply:
I agree that they are always in the same cache line (it is really always, not just most of the time, because we over align to `alignas(2 * sizeof(void*))`). So from the cache perspective it does not matter.

I agree also that eliminating debug mode call can be potentially beneficial.

Still I more like the original order, because we compare `_Rep == _Old_rep;` the very last thing, as close to wait as possible, so reduce the probability of resorting to timeout.

Reply:
Considering _Rep is just a stack-local and the real load is on the line above, I can't quite follow your reasoning.

Reply:
> Considering _Rep is just a stack-local and the real load is on the line above, I can't quite follow your reasoning.

Hm, you are right, we done loading it into memory at the same time anyway.

Then debug mode call saving could be the main decision factor here.

Reply:
Ah, I'd forgotten that we overalign - my cache coherence point is invalid. +1 to avoiding the call in debug compiles, though. We _really_ want to minimize the time between load and check.